LoopUI/Charts/ChartsManager.swift

```diff
@@ -0,0 +1,246 @@
+            chartsCache[index] = charts[index].generate(withFrame: frame, xAxisModel: xAxisModel, xAxisValues: xAxisValues, axisLabelSettings: axisLabelSettings, guideLinesLayerSettings: guideLinesLayerSettings, colors: colors, chartSettings: chartSettings, labelsWidthY: labelsWidthY, gestureRecognizer: gestureRecognizer)
+        }
+
+        return chartsCache[index]
+    }
+
+    public func invalidateChart(atIndex index: Int) {
+        chartsCache[index] = nil
+    }
+
+    // MARK: - Shared Axis
+
+    private func generateXAxisValues() {
+        if let endDate = charts.compactMap({ $0.endDate }).max() {
+            updateEndDate(endDate)
+        }
+
+        let timeFormatter = DateFormatter()
+        timeFormatter.dateFormat = "h a"
+
+        let points = [
+            ChartPoint(
+                x: ChartAxisValueDate(date: startDate, formatter: timeFormatter),
+                y: ChartAxisValue(scalar: 0)
+            ),
+            ChartPoint(
+                x: ChartAxisValueDate(date: endDate, formatter: timeFormatter),
+                y: ChartAxisValue(scalar: 0)
+            )
+        ]
+
+        let segments = ceil(endDate.timeIntervalSince(startDate).hours)
+
+        let xAxisValues = ChartAxisValuesStaticGenerator.generateXAxisValuesWithChartPoints(points,
+            minSegmentCount: segments - 1,
+            maxSegmentCount: segments + 1,
+            multiple: TimeInterval(hours: 1),
+            axisValueGenerator: {
+                ChartAxisValueDate(
+                    date: ChartAxisValueDate.dateFromScalar($0),
+                    formatter: timeFormatter,
+                    labelSettings: self.axisLabelSettings
+                )
+            },
+            addPaddingSegmentIfEdge: false
+        )
+        xAxisValues.first?.hidden = true
+        xAxisValues.last?.hidden = true
+
+        self.xAxisValues = xAxisValues
+    }
+
+    /// Runs any necessary steps before rendering charts
+    public func prerender() {
+        if xAxisValues == nil {
+            generateXAxisValues()
+        }
+    }
+}
+
+fileprivate extension Array {
+    mutating func replaceAllElements(with element: Element) {
+        self = Array(repeating: element, count: count)
+    }
+}
+
+
+public protocol ChartProviding {
+    /// Instructs the chart to clear its non-critical resources like caches
+    func didReceiveMemoryWarning()
+
+    /// The last date represented in the chart data
+    var endDate: Date? { get }
+
+    /// Creates a chart from the current data
+    ///
+    /// - Parameters:
+    ///   - frame:
+    ///   - xAxisModel:
+    ///   - xAxisValues:
+    ///   - axisLabelSettings:
+    ///   - guideLinesLayerSettings:
+    ///   - colors:
+    ///   - chartSettings:
+    ///   - labelsWidthY:
+    ///   - gestureRecognizer:
+    /// - Returns: A new chart object
+    func generate(withFrame frame: CGRect,
```

Comment:
Empty parameter docs should either be filled out or removed

Reply:
Removing for now.